### PR TITLE
feat(discord): update discord typings

### PR DESCRIPTION
**Pull request by XxLittleCxX**

### CODE REVIEW

File: adapters/discord/src/types/component.ts [hunk 1]
```diff
@@ -1,32 +1,6 @@
...
-  /** custom placeholder text if nothing is selected, max 100 characters */
-  placeholder?: string
-  /** the minimum number of items that must be chosen; default 1, min 0, max 25 */
-  min_values?: integer
-  /** the maximum number of items that can be chosen; default 1, max 25 */
-  max_values?: integer
-  /** a list of child components */
-  components?: Component[]
-}
+export type Component = Button | SelectMenu | TextInput
```
Comment (shigma) [inline adapters/discord/src/types/component.ts:3]: There are 8 types mentioned below, so why are only 3 discriminants here?

File: adapters/discord/src/types/application.ts [hunk 2]
```diff
@@ -22,7 +22,8 @@ export interface Application {
   privacy_policy_url?: string
   /** partial user object containing info on the owner of the application */
   owner?: Partial<User>
-  /** if this application is a game sold on Discord, this field will be the summary field for the store page of its primary sku */
+  /** deprecated, if this application is a game sold on Discord, this field will be the summary field for the store page of its primary sku
+  */
```
Comment (shigma) [inline adapters/discord/src/types/application.ts:26]:
```suggestion
  /** deprecated, if this application is a game sold on Discord, this field will be the summary field for the store page of its primary sku */
```

File: adapters/discord/src/types/application.ts [hunk 3]
```diff
@@ -38,6 +39,21 @@ export interface Application {
   cover_image?: string
   /** the application's public flags */
   flags?: integer
+  /** up to 5 tags describing the content and functionality of the application */
+  tags?: [string, string?, string?, string?, string?]
+  /** settings for the application's default in-app authorization link, if enabled */
+  install_params?: InstallParams
+  /**	the application's default custom authorization link, if enabled */
```
Comment (shigma) [inline adapters/discord/src/types/application.ts:46]:
```suggestion
  /** the application's default custom authorization link, if enabled */
```

Replace tabs with spaces (with more occurences below)

File: adapters/discord/src/types/component.ts (hunk 1 above)
Comment (shigma) [inline adapters/discord/src/types/component.ts:3]: There are 8 component types below, so why are only 3 discriminants here?

File: adapters/discord/src/types/gateway.ts [hunk 4]
```diff
@@ -46,7 +46,7 @@ export enum GatewayOpcode {
   HEARTBEAT_ACK = 11,
 }
 
-/** https://discord.com/developers/docs/topics/gateway#gateway-intents */
+/** https://discord.com/developers/docs/topics/gateway-events#gateway-intents */
```
Comment (shigma) [inline adapters/discord/src/types/gateway.ts:49]: Obviously this should be https://discord.com/developers/docs/topics/gateway#gateway-intents.

It seems that you have done some replacements incorrectly.

File: adapters/discord/src/types/gateway.ts [hunk 5]
```diff
@@ -216,7 +216,7 @@ export interface IdentifyParams {
   intents: integer
 }
 
-/** https://discord.com/developers/docs/topics/gateway-events#identify-identify-connection-properties */
+/** https://discord.com/developers/docs/topics/gateway-events-events#identify-identify-connection-properties */
```
Comment (shigma) [inline adapters/discord/src/types/gateway.ts:219]: 404

File: adapters/discord/src/types/interaction.ts [hunk 6]
```diff
@@ -32,49 +32,71 @@ export interface Interaction {
...
+    /** 	values the user selected in a select menu component */
+    values?: string[]
+  }
+
+  /** https://discord.com/developers/docs/interactions/receiving-and-responding#interaction-object-modal-submit-data-structure */
+  export interface ModalSubmit {
+    custom_id: string
+    components: Component[]
+  }
+
```
Comment (shigma) [inline adapters/discord/src/types/interaction.ts:74]:
```suggestion
```

**Review by shigma — CHANGES_REQUESTED** [review]

**Comment by shigma** [comment]
Also you need to update the date in types/index.ts

File: adapters/discord/src/types/component.ts (hunk 1 above)
Comment (XxLittleCxX) [reply adapters/discord/src/types/component.ts]: `SelectMenu` seems to contain the other four

https://discord.com/developers/docs/interactions/message-components#select-menu-types

File: adapters/discord/src/types/component.ts [hunk 7]
```diff
@@ -22,6 +22,12 @@ export enum ComponentType {
   CHANNEL_SELECT = 8
 }
 
+/** https://discord.com/developers/docs/interactions/message-components#action-rows */
+export interface ActionRow {
+  type: ComponentType.ACTION_ROW,
```
Comment (shigma) [inline adapters/discord/src/types/component.ts:27]:
```suggestion
  type: ComponentType.ACTION_ROW
```

Useless comma (idk if there are more)

File: adapters/discord/src/types/application-role-connection.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
+import { Internal, Locale } from "."
```
Comment (shigma) [inline adapters/discord/src/types/application-role-connection.ts:1]:
```suggestion
import { Internal, Locale } from '.'
```

File: adapters/discord/src/types/application-role-connection.ts [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  /** https://discord.com/developers/docs/resources/application-role-connection-metadata#application-role-connection-metadata-object-application-role-connection-metadata-structure */
+  export interface Metadata {
+    /** type of metadata value */
+    type: MetadataType
+    /** dictionary key for the metadata field (must be a-z, 0-9, or _ characters; 1-50 characters) */
+    key: string
+    /** name of the metadata field (1-100 characters) */
+    name: string
+    /** translations of the name */
+    name_localizations?: Record<Locale, 'string'>
```
Comment (shigma) [inline adapters/discord/src/types/application-role-connection.ts:13]:
```suggestion
    name_localizations?: Record<Locale, string>
```

这里肯定不是 `'string'` 吧？

File: adapters/discord/src/types/application-role-connection.ts [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    /** description of the metadata field (1-200 characters) */
+    description: string
+    /** translations of the description */
+    description_localizations?: Record<Locale, 'string'>
```
Comment (shigma) [inline adapters/discord/src/types/application-role-connection.ts:17]:
```suggestion
    description_localizations?: Record<Locale, string>
```

File: adapters/discord/src/types/application.ts [hunk 11]
```diff
@@ -48,9 +63,12 @@ export enum ApplicationFlag {
   GATEWAY_GUILD_MEMBERS_LIMITED = 1 << 15,
   VERIFICATION_PENDING_GUILD_LIMIT = 1 << 16,
   EMBEDDED = 1 << 17,
+  GATEWAY_MESSAGE_CONTENT = 1 << 18,
+  GATEWAY_MESSAGE_CONTENT_LIMITED = 1 << 19,
+  APPLICATION_COMMAND_BADGE = 1 << 23
```
Comment (shigma) [inline adapters/discord/src/types/application.ts:68]:
```suggestion
  APPLICATION_COMMAND_BADGE = 1 << 23,
```

File: adapters/discord/src/types/auto-moderation.ts [hunk 12]
```diff
@@ -59,6 +59,8 @@ export namespace AutoModerationRule {
   export interface TriggerMetadata {
     /** associated with `KEYWORD`: substrings which will be searched for in content */
     keyword_filter: string[]
+    /** 	regular expression patterns which will be matched against content (Maximum of 10) */
```
Comment (shigma) [inline adapters/discord/src/types/auto-moderation.ts:62]:
```suggestion
    /** regular expression patterns which will be matched against content (Maximum of 10) */
```

File: adapters/discord/src/types/command.ts [hunk 13]
```diff
@@ -79,6 +91,7 @@ export namespace ApplicationCommand {
     MENTIONABLE = 9,
     /** Any double between -2^53 and 2^53 */
     NUMBER = 10,
+    ATTACHMENT = 11
```
Comment (shigma) [inline adapters/discord/src/types/command.ts:94]:
```suggestion
    /** attachment object */
    ATTACHMENT = 11,
```

File: adapters/discord/src/types/command.ts [hunk 14]
```diff
@@ -128,26 +143,46 @@ export namespace ApplicationCommand {
     export interface Create {
       /** 1-32 character name */
       name: string
+      /**	localization dictionary for the name field. Values follow the same restrictions as name */
```
Comment (shigma) [inline adapters/discord/src/types/command.ts:146]: 这下面还一堆 tab

File: adapters/discord/src/types/component.ts [hunk 15]
```diff
@@ -1,32 +1,6 @@
-import { Emoji, integer } from '.'
+import { Emoji, integer, Channel } from '.'
```
Comment (shigma) [inline adapters/discord/src/types/component.ts:1]:
```suggestion
import { Channel, Emoji, integer } from '.'
```

File: adapters/discord/src/types/component.ts [hunk 16]
```diff
@@ -38,14 +12,28 @@ export enum ComponentType {
...
   /** A text input object */
   TEXT_INPUT = 4,
+  /** Select menu for users */
+  USER_SELECT = 5,
+  /** Select menu for roles */
+  ROLE_SELECT = 6,
+  /** Select menu for mentionables (users and roles) */
+  MENTIONABLE_SELECT = 7,
+  /** Select menu for channels */
+  CHANNEL_SELECT = 8
```
Comment (shigma) [inline adapters/discord/src/types/component.ts:22]:
```suggestion
  CHANNEL_SELECT = 8,
```

File: adapters/discord/src/types/guild.ts [hunk 17]
```diff
@@ -98,10 +80,10 @@ export interface Guild {
   welcome_screen?: WelcomeScreen
   /** guild NSFW level */
   nsfw_level: integer
-  /** Stage instances in the guild */
-  stage_instances?: StageInstance[]
   /** custom guild stickers */
   stickers?: Sticker[]
+  /**	whether the guild has the boost progress bar enabled */
```
Comment (shigma) [inline adapters/discord/src/types/guild.ts:85]: tab，下面还有

File: adapters/discord/src/types/integration.ts [hunk 18]
```diff
@@ -32,6 +32,8 @@ export interface Integration {
   revoked?: boolean
   /** The bot/OAuth2 application for discord integrations */
   application?: IntegrationApplication
+  /**	the scopes the application has been authorized for */
```
Comment (shigma) [inline adapters/discord/src/types/integration.ts:35]: tab

File: adapters/discord/src/types/interaction.ts [hunk 19]
```diff
@@ -32,49 +32,70 @@ export interface Interaction {
...
+  }
+}
+
 /** https://discord.com/developers/docs/interactions/receiving-and-responding#interaction-object-interaction-type */
 export enum InteractionType {
   PING = 1,
   APPLICATION_COMMAND = 2,
   MESSAGE_COMPONENT = 3,
+  APPLICATION_COMMAND_AUTOCOMPLETE = 4,
+  MODAL_SUBMIT = 5
```
Comment (shigma) [inline adapters/discord/src/types/interaction.ts:82]:
```suggestion
  MODAL_SUBMIT = 5,
```

**Review by shigma — APPROVED** [review]

**Pull request merged**